aten/src/ATen/test/cuda_half_test.cu

```diff
@@ -0,0 +1,78 @@
+#define CATCH_CONFIG_MAIN
+#include "catch.hpp"
+
+#include "ATen/ATen.h"
+#include "ATen/cuda/CUDANumerics.cuh"
+#include "cuda.h"
+#include "cuda_fp16.h"
+#include "cuda_runtime.h"
+
+#include <assert.h>
+
+using namespace at;
+
+__host__ __device__ void test(){
+  
+  // test half construction and implicit conversions in device
+  assert(Half(3) == Half(3.0f));
+  assert(static_cast<Half>(3.0f) == Half(3.0f));
+  // there is no float <=> __half implicit conversion
+  assert(static_cast<Half>(3.0f) == 3.0f);
+
+  // asserting if the  functions used on 
+  // half types give almost equivalent results when using
+  //  functions on double.
+  // The purpose of these asserts are to test the device side
+  // half API for the common mathematical functions.
+  // Note: When calling std math functions from device, don't
+  // use the std namespace, but just "::" so that the function
+  // gets resolved from nvcc math_functions.hpp
```

Comment:
What happens if you use the std:: qualifier?

Reply:
I think I was getting errors in the CI which says something like `__host__` function called from `__host__ __device__` is not allowed and then a following trace of a __builtin* function called with the corresponding `std::*` function. I assumed that when using std:: qualifier in __device__, it's resolving from `cmath` which calls `__builtin*` vs, resolving from nvcc header `/usr/local/cuda/include/crt/math_functions.hpp` which marks the std:: functions with `__host__ __device__`.